crates/ursa-rpc-service/src/api.rs

```diff
@@ -103,41 +104,41 @@ impl<S> NetworkInterface for NodeNetworkInterface<S>
 where
     S: Blockstore + Store + Send + Sync + 'static,
 {
-    async fn get(&self, cid: Cid) -> Result<Option<Vec<u8>>> {
+    async fn check_blockstore(&self, cid: Cid) -> Result<()> {
         if !self.store.blockstore().has(&cid)? {
             info!("Requesting block with the cid {cid:?}");
             let (sender, receiver) = oneshot::channel();
             let request = NetworkCommand::GetBitswap { cid, sender };
 
             // use network sender to send command
             self.network_send.send(request)?;
+
             if let Err(e) = receiver.await? {
```

Comment:
It doesn't look like it's happening concurrently if we're waiting on this receiver? Am I reading this right?

Reply:
All open receivers here are triggered by any successful response for a given cid be it bitswap or origin

Reply:
But we havent made the origin request yet at this point, no? if this returns an error, it sends the GetOriging command, or is the GET origin request triggered somewhere before this line?

Reply:
oh, i understand what you're saying now. Basically flow is this:
1. attempt bitswap
2. bitswap failed attempt origin

But concurrent users requesting the same content would resolve from a single request no matter what step they are at

Reply:
It does bring up a point though, where we could request bitswap and origin at the same time and use whichever responds first

Reply:
That's what I thought you meant with concurrent requests. 

> The goal is to share successful responses for concurrent requests for cids from both bitswap and origin, so that latency is reduced for new content, as well as retain the exact same flow of using bitswap or the origin'

Reply:
> But concurrent users requesting the same content would resolve from a single request no matter what step they are at

In that case a new behavior isn't really necessary.

Reply:
https://github.com/fleek-network/ursa/pull/278/commits/11e9610901aad513e46cbede8f05907d30ba1363